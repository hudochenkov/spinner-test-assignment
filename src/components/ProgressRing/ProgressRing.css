.progress-ring {
	position: relative;
}
.progress-ring__image {
}
.progress-ring--active .progress-ring__image {
	transform-origin: 50%;
	animation: progress-ring-spin 2s linear infinite;
}
.progress-ring__background-ring {
	stroke: #eee;
}
.progress-ring__arc {
	transition: stroke-dashoffset 0.2s;
	stroke: #42a1ff;
	stroke-linecap: round;
}
.progress-ring__label {
	position: absolute;
	top: 0;
	bottom: 0;
	display: flex;
	align-items: center;
	justify-content: center;
	width: 100%;
	height: 100%;
}
.progress-ring__label-itself {
	padding-left: 6px; /* For nice optical centering of a text */
	font-size: 34px;
}
.progress-ring__label-itself:after {
	color: #aaa;
	content: "%";
	font-size: 16px;
}

@keyframes progress-ring-spin {
	from {
		transform: rotate(0deg);
	}

	to {
		transform: rotate(360deg);
	}
}
