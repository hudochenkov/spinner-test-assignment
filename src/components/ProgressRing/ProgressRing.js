import React from 'react';
import PropTypes from 'prop-types';
import './ProgressRing.css';

function ProgressRing(props) {
	// Round in case it's not an integer
	const value = Math.round(props.value);

	const size = 160;
	const strokeWidth = 14;

	const centerPoint = size / 2;
	const radius = (size - strokeWidth) / 2;

	const circumference = (radius * 2 * Math.PI).toFixed(3);
	const progressLine = (circumference - (value / 100) * circumference).toFixed(3);

	// Using inline styles, because size potentially could be configurable via props
	// Don't apply animations (.progress-ring--active) when running animation won't be visible
	return (
		<div
			className={`progress-ring ${props.isActive ? 'progress-ring--active' : ''}`}
			style={{ width: size, height: size }}
		>
			<svg
				className="progress-ring__image"
				width={size}
				height={size}
				viewBox={`0 0 ${size} ${size}`}
				xmlns="http://www.w3.org/2000/svg"
			>
				<circle
					className="progress-ring__background-ring"
					strokeWidth={strokeWidth}
					fill="transparent"
					r={radius}
					cx={centerPoint}
					cy={centerPoint}
				/>

				<circle
					className="progress-ring__arc"
					strokeWidth={strokeWidth}
					fill="transparent"
					r={radius}
					cx={centerPoint}
					cy={centerPoint}
					style={{
						strokeDasharray: `${circumference} ${circumference}`,
						strokeDashoffset: progressLine,
					}}
				/>
			</svg>

			<div className="progress-ring__label">
				<span className="progress-ring__label-itself">{value}</span>
			</div>
		</div>
	);
}

ProgressRing.propTypes = {
	isActive: PropTypes.bool,
	value: PropTypes.number,
};

ProgressRing.defaultProps = {
	isActive: false,
	value: 0,
};

export default ProgressRing;
