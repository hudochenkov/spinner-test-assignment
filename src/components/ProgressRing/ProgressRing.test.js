import React from 'react';
import { shallow } from 'enzyme';
import ProgressRing from './ProgressRing';

describe('<ProgressRing />', () => {
	it('without props', () => {
		const wrapper = shallow(<ProgressRing />);

		expect(wrapper).toMatchSnapshot();
	});

	it('with a value', () => {
		const wrapper = shallow(<ProgressRing value={15} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('with a value and active', () => {
		const wrapper = shallow(<ProgressRing value={15} isActive />);

		expect(wrapper).toMatchSnapshot();
	});
});
