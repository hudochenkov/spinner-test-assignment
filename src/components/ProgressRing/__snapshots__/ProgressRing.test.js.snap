// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressRing /> with a value 1`] = `
<div
  className="progress-ring "
  style={
    Object {
      "height": 160,
      "width": 160,
    }
  }
>
  <svg
    className="progress-ring__image"
    height={160}
    viewBox="0 0 160 160"
    width={160}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="progress-ring__background-ring"
      cx={80}
      cy={80}
      fill="transparent"
      r={73}
      strokeWidth={14}
    />
    <circle
      className="progress-ring__arc"
      cx={80}
      cy={80}
      fill="transparent"
      r={73}
      strokeWidth={14}
      style={
        Object {
          "strokeDasharray": "458.673 458.673",
          "strokeDashoffset": "389.872",
        }
      }
    />
  </svg>
  <div
    className="progress-ring__label"
  >
    <span
      className="progress-ring__label-itself"
    >
      15
    </span>
  </div>
</div>
`;

exports[`<ProgressRing /> with a value and active 1`] = `
<div
  className="progress-ring progress-ring--active"
  style={
    Object {
      "height": 160,
      "width": 160,
    }
  }
>
  <svg
    className="progress-ring__image"
    height={160}
    viewBox="0 0 160 160"
    width={160}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="progress-ring__background-ring"
      cx={80}
      cy={80}
      fill="transparent"
      r={73}
      strokeWidth={14}
    />
    <circle
      className="progress-ring__arc"
      cx={80}
      cy={80}
      fill="transparent"
      r={73}
      strokeWidth={14}
      style={
        Object {
          "strokeDasharray": "458.673 458.673",
          "strokeDashoffset": "389.872",
        }
      }
    />
  </svg>
  <div
    className="progress-ring__label"
  >
    <span
      className="progress-ring__label-itself"
    >
      15
    </span>
  </div>
</div>
`;

exports[`<ProgressRing /> without props 1`] = `
<div
  className="progress-ring "
  style={
    Object {
      "height": 160,
      "width": 160,
    }
  }
>
  <svg
    className="progress-ring__image"
    height={160}
    viewBox="0 0 160 160"
    width={160}
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="progress-ring__background-ring"
      cx={80}
      cy={80}
      fill="transparent"
      r={73}
      strokeWidth={14}
    />
    <circle
      className="progress-ring__arc"
      cx={80}
      cy={80}
      fill="transparent"
      r={73}
      strokeWidth={14}
      style={
        Object {
          "strokeDasharray": "458.673 458.673",
          "strokeDashoffset": "458.673",
        }
      }
    />
  </svg>
  <div
    className="progress-ring__label"
  >
    <span
      className="progress-ring__label-itself"
    >
      0
    </span>
  </div>
</div>
`;
